File: Data Structures & Algorithms/pow-x-n/submission-1.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        ans = self.binaryExp(x, abs(n))
        return ans if n >= 0 else 1/ans
    
    def binaryExp(self, x: float, n: int) -> float:
        if x == 0:
            return 0
        if n == 0:
            return 1

        ans = self.binaryExp(x * x, n // 2)
        if n % 2:
            ans = x * ans

        return ans

File: Data Structures & Algorithms/last-stone-weight/submission-1.js
/**
 * const { MaxPriorityQueue } = require('@datastructures-js/priority-queue');
 */

class Solution {
    /**
     * @param {number[]} stones
     * @return {number}
     */
    lastStoneWeight(stones) {
        const maxHeap = new MaxPriorityQueue()

        for (let num of stones) {
            maxHeap.enqueue(num)
        }

        while (maxHeap.size() > 1) {
            let x = maxHeap.dequeue()
            let y = maxHeap.dequeue()
            maxHeap.enqueue(Math.abs(x - y))
        }

        return maxHeap.size() === 1 ? maxHeap.dequeue() : 0
    }
}
